### PR TITLE
Error when running CodeCompass binary from extended PATH

### LINKED ISSUE #353: Error when running CodeCompass binary from extended PATH
1. I built a CodeCompass under `/home/username/CodeCompass-build`.
2. I installed it under `/home/username/CodeCompass-install`.
3. I extended my PATH environment variable:
`extend PATH="/home/username/CodeCompass-install/bin:$PATH"`.

When I try to run for example the `CodeCompass_webserver -h` command from `/home/username/CodeCompass-build` directory I will get the following error message:
```sh
terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
  what():  boost::filesystem::canonical: No such file or directory: "/home/username/CodeCompass-build/CodeCompass_webserver"
Aborted (core dumped
```

This error comes from this code:
https://github.com/Ericsson/CodeCompass/blob/ebbe021cbcfb78b4e4084c34c1eb46a4e2cbdc08/util/src/filesystem.cpp#L18-L19

```
path -> CodeCompass_webserver
fs::path(path)) -> "CodeCompass_webserver"
fs::system_complete(fs::path(path)) -> "/home/username/CodeCompass-build/CodeCompass_webserver"
```
So it will not work if I am not in the installed bin directory.

### PR DESCRIPTION
**Pull request by intjftw**
Webserver path is searched for in PATH if not called from its install directory.
Fixes #353.

### CODE REVIEW

File: util/src/filesystem.cpp [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #include <boost/filesystem.hpp>
 
 #include <util/filesystem.h>
+#include <util/logutil.h>
+
+#include <stdlib.h>
```
Comment (mcserep) [inline util/src/filesystem.cpp]: `stdlib.h` is a C header. Use the C++ version: `cstdlib`

File: util/src/filesystem.cpp [hunk 2]
```diff
@@ -1,6 +1,9 @@
 #include <boost/filesystem.hpp>
 
 #include <util/filesystem.h>
+#include <util/logutil.h>
```
Comment (mcserep) [inline util/src/filesystem.cpp]: I don't think this is needed.

File: util/src/filesystem.cpp [hunk 3]
```diff
@@ -15,8 +18,34 @@ std::string binaryPathToInstallDir(const char* path)
...
+
+      if (fs::exists(fs::path(webserverPath)))
+      {
+        return fs::canonical(fs::path(webserverPath))
+          .parent_path().parent_path().string();
+      }
+
+      fullPath.erase(0, pos + delimiter.length());
+    }
+  }
```
Comment (mcserep) [inline util/src/filesystem.cpp]: If no proper prefix is found in `PATH`, then there is no value returned from the function. Although it shall *never* occur according to the program workflow, an undefined behavior like that should still be avoided. Throwing an exception (e.g. `runtime_error`) could be a solution.

File: util/src/filesystem.cpp [hunk 4]
```diff
@@ -15,8 +18,34 @@ std::string binaryPathToInstallDir(const char* path)
...
+  // environment variable.
+
+  if (fs::exists(fs::system_complete(fs::path(path))))
+  {
+    return fs::canonical(fs::system_complete(fs::path(path)))
+      .parent_path().parent_path().string();
+  }
+  else
+  {
+    std::string webserverPath;
```
Comment (csordasmarton) [inline util/src/filesystem.cpp]: This variable name is a little bit misleading. We can use this function to search not only webserver binary. So maybe a different name would be better, for example `pPath`.

**Comment by intjftw** [comment]
@csordasmarton The proposed search_path() function is in the Boost.Process library which was [introduced](https://www.boost.org/users/history/version_1_64_0.html) in Boost version 1.64. The standard Ubuntu package repository for Xenial contains 1.58.

File: util/src/filesystem.cpp [hunk 5]
```diff
@@ -15,8 +17,36 @@ std::string binaryPathToInstallDir(const char* path)
...
+  else
+  {
+    std::string pPath;
+    std::string fullPath = std::getenv("PATH");
+    std::string delimiter = ":";
+
+    std::size_t pos = 0;
+    while ((pos = fullPath.find(delimiter)) != std::string::npos)
+    {
+      pPath = fullPath.substr(0, pos) + "/" + path;
```
Comment (bruntib) [inline util/src/filesystem.cpp]: `fs::path` has an `operator/` overload for joining paths which uses the delimiter of the current platform. We should use this for joining paths.

File: util/src/filesystem.cpp [hunk 6]
```diff
@@ -15,8 +17,36 @@ std::string binaryPathToInstallDir(const char* path)
...
+      .parent_path().parent_path().string();
+  }
+  else
+  {
+    std::string pPath;
+    std::string fullPath = std::getenv("PATH");
+    std::string delimiter = ":";
+
+    std::size_t pos = 0;
+    while ((pos = fullPath.find(delimiter)) != std::string::npos)
```
Comment (bruntib) [inline util/src/filesystem.cpp]: `std::strtok()` would be more convenient here.

File: util/src/filesystem.cpp [hunk 7]
```diff
@@ -15,8 +17,36 @@ std::string binaryPathToInstallDir(const char* path)
   // of any symbolic links and relative references.
   // The path is then parented twice (from binary to "bin" folder, and then
   // the root).
-  return fs::canonical(fs::system_complete(fs::path(path)))
-    .parent_path().parent_path().string();
+  // If the path does not exist, the given binary is searched for in the PATH
+  // environment variable.
```
Comment (bruntib) [inline util/src/filesystem.cpp]: We could write a TODO here to use `search_path()` from Boost in the future.

File: util/src/filesystem.cpp (hunk 5 above)
Comment (bruntib) [reply util/src/filesystem.cpp]: By the way the separator in `$PATH` can also be queried in some languages. I don't know if this is available in Boost, that's why I havet't requested a change here. But for example Windows uses `;` and unix systems use `:`. Ok, I know we don't support Windows... :)

**Pull request merged**